src/substrait/function_registry.py

```diff
@@ -0,0 +1,316 @@
+        if parameter_name == "any":
+            return True
+        else:
+            if parameter_name in parameters and parameters[
+                parameter_name
+            ].SerializeToString(deterministic=True) != dtype.SerializeToString(
+                deterministic=True
+            ):
+                return False
+            else:
+                parameters[parameter_name] = dtype
+                return True
+
+    kind = dtype.WhichOneof("kind")
+
+    if kind != expected_kind:
+        return False
+
+    if kind == "decimal":
+        if violates_integer_option(
+            dtype.decimal.scale, parameterized_type.decimal.scale, parameters
+        ) or violates_integer_option(
+            dtype.decimal.precision, parameterized_type.decimal.precision, parameters
+        ):
+            return False
+
+    # TODO handle all types
+
+    return True
+
+
+class FunctionEntry:
+    def __init__(self, uri: str, name: str, impl: Mapping[str, Any]) -> None:
+        self.name = name
+        self.normalized_inputs: list = []
+        self.uri: str = uri
+        self.anchor = next(id_generator)
+        self.arguments = []
+        self.rtn = impl["return"]
+        self.nullability = impl.get("nullability", False)
+        self.variadic = impl.get("variadic", False)
+        if input_args := impl.get("args", []):
+            for val in input_args:
+                if typ := val.get("value"):
+                    self.arguments.append(to_parameterized_type(typ.strip("?")))
+                    self.normalized_inputs.append(normalize_substrait_type_names(typ))
+                elif arg_name := val.get("name", None):
+                    self.arguments.append(val.get("options"))
+                    self.normalized_inputs.append("req")
+
+    def __repr__(self) -> str:
+        return f"{self.name}:{'_'.join(self.normalized_inputs)}"
+
+    def satisfies_signature(self, signature: tuple) -> Optional[str]:
+        if self.variadic:
+            min_args_allowed = self.variadic.get("min", 0)
+            if len(signature) < min_args_allowed:
+                return None
+            inputs = [self.arguments[0]] * len(signature)
+        else:
+            inputs = self.arguments
+        if len(inputs) != len(signature):
+            return None
+
+        zipped_args = list(zip(inputs, signature))
+
+        parameters = {}
+
+        for x, y in zipped_args:
+            if type(y) == str:
+                if y not in x:
+                    return None
+            else:
+                if not covers(y, x, parameters):
+                    return None
+
+        return evaluate(self.rtn, parameters)
+
+
+class FunctionRegistry:
+    def __init__(self) -> None:
+        self._function_mapping: dict = defaultdict(dict)
+        self.id_generator = itertools.count(1)
+
+        self.uri_aliases = {}
+
+        for fpath in importlib_files("substrait.extensions").glob(  # type: ignore
+            "functions*.yaml"
+        ):
```

Comment:
Consider making a constant for the default URI prefix.

Reply:
done